cryptarchia/cryptarchia.py

```diff
@@ -696,45 +646,99 @@ def _is_slot_leader(self, epoch: EpochState, slot: Slot):
         )
 
 
-def common_prefix_len(a: Chain, b: Chain) -> int:
-    for i, (x, y) in enumerate(zip(a.blocks, b.blocks)):
-        if x.id() != y.id():
-            return i
-    return min(len(a.blocks), len(b.blocks))
+def iter_chain(tip: Id, states: Dict[Id, LedgerState]):
+    while tip in states:
+        yield states[tip]
+        tip = states[tip].block.parent
+
+
+def chain_suffix(tip: Id, n: int, states: Dict[Id, LedgerState]) -> list[LedgerState]:
+    return list(reversed(list(itertools.islice(iter_chain(tip, states), n))))
+
+
+def common_prefix_depth(a: Id, b: Id, states: Dict[Id, LedgerState]) -> (int, int):
+    a_blocks = iter_chain(a, states)
+    b_blocks = iter_chain(b, states)
+
+    seen = {}
+    depth = 0
+    while True:
+        try:
+            a_block = next(a_blocks).block.id()
+            if a_block in seen:
+                # we had seen this block from the fork chain
+                return depth, seen[a_block]
 
+            seen[a_block] = depth
+        except StopIteration:
+            pass
 
-def chain_density(chain: Chain, slot: Slot) -> int:
-    return len(
-        [
-            block
-            for block in chain.blocks
-            if block.slot.absolute_slot < slot.absolute_slot
-        ]
-    )
+        try:
+            b_block = next(b_blocks).block.id()
+            if b_block in seen:
+                # we had seen the fork in the local chain
+                return seen[b_block], depth
+            seen[b_block] = depth
+        except StopIteration:
+            pass
```

Comment:
Btw, if you trim the longest iterator first, you can then proceed in tandem without having to store intermediate blocks